src/addr.rs

```diff
@@ -1,335 +1,194 @@
+
+            /// Aligns the address downwards to 4096 (bytes).
+            #[inline]
+            pub const fn align_down_4k(self) -> Self {
+                Self::from_usize(align_down(self.0, PAGE_SIZE_4K))
+            }
+
+            /// Aligns the address upwards to 4096 (bytes).
+            #[inline]
+            pub const fn align_up_4k(self) -> Self {
+                Self::from_usize(align_up(self.0, PAGE_SIZE_4K))
+            }
+
+            /// Returns the offset of the address within a 4K-sized page.
+            #[inline]
+            pub const fn align_offset_4k(self) -> usize {
+                align_offset(self.0, PAGE_SIZE_4K)
+            }
+
+            /// Checks whether the address is 4K-aligned.
+            #[inline]
+            pub const fn is_aligned_4k(self) -> bool {
+                is_aligned(self.0, PAGE_SIZE_4K)
+            }
+        }
+
+        impl From<usize> for $name {
+            #[inline]
+            fn from(addr: usize) -> Self {
+                Self(addr)
+            }
+        }
+
+        impl From<$name> for usize {
+            #[inline]
+            fn from(addr: $name) -> usize {
+                addr.0
+            }
+        }
+
+        impl core::ops::Add<usize> for $name {
+            type Output = Self;
+            #[inline]
+            fn add(self, rhs: usize) -> Self {
+                Self(self.0 + rhs)
+            }
+        }
+
+        impl core::ops::AddAssign<usize> for $name {
+            #[inline]
+            fn add_assign(&mut self, rhs: usize) {
+                self.0 += rhs;
+            }
+        }
+
+        impl core::ops::Sub<usize> for $name {
+            type Output = Self;
+            #[inline]
+            fn sub(self, rhs: usize) -> Self {
+                Self(self.0 - rhs)
+            }
+        }
+
+        impl core::ops::SubAssign<usize> for $name {
+            #[inline]
+            fn sub_assign(&mut self, rhs: usize) {
+                self.0 -= rhs;
+            }
+        }
+
+        impl core::fmt::Debug for $name {
+            fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+                f.write_fmt(format_args!(concat!($display_prefix, "{:#x}"), self.0))
+            }
+        }
+
+        impl core::fmt::LowerHex for $name {
+            fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+                f.write_fmt(format_args!(concat!($display_prefix, "{:#x}"), self.0))
+            }
+        }
+
+        impl core::fmt::UpperHex for $name {
+            fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+                f.write_fmt(format_args!(concat!($display_prefix, "{:#X}"), self.0))
+            }
+        }
+    };
 }
 
-impl fmt::UpperHex for VirtAddr {
-    #[inline]
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        f.write_fmt(format_args!("VA:{:#X}", self.0))
-    }
-}
+def_addr_type!(PhysAddr, "A physical memory address.", "PA:");
```

Comment:
I think the macro API should be like this:
```rust
def_addr! {
  pub struct PhysAddr = usize;
  pub(crate) struct AddrU64 = u64;
  struct AddrU32 = u32; 
}
```

Reply:
How about this?

```rust
def_addr! {
    #[doc = "A physical memory address."]
    pub struct PhysAddr(usize);
    pub(crate) struct AddrU64(u64);
    struct AddrU32(u32);
}
```

For consistency with rust `struct` definition.

Reply:
Btw what should `From::<usize>::from` and `Into::<usize>::into` do if the inner type is shorter than `usize` (like `u32`)?

Reply:
> How about this?
> 
> 
> 
> ```rust
> 
> def_addr! {
> 
>     #[doc = "A physical memory address."]
> 
>     pub struct PhysAddr(usize);
> 
>     pub(crate) struct AddrU64(u64);
> 
>     struct AddrU32(u32);
> 
> }
> 
> ```
> 
> 
> 
> For consistency with rust `struct` definition.

You can still use `/// xxxx` for comments.

Reply:
> You can still use `/// xxxx` for comments.

Yes😂 just found this.